### ISSUE TITLE
Enhance supervisor

### ISSUE THREAD

**Issue by jdegoes** [zio/zio-actors#15]
The supervisor takes a schedule, but could also take an "or else" function to do if the retry fails; see `IO#retryOrElse0` for inspiration on what this could look like.

**Comment by mijicd** [comment]
@jdegoes I need your advice here. 

Namely, I've tried to change the encoding of the `Supervisor` as follows:

```scala
trait Supervisor[-E] {
  type A0

  def supervise(io: IO[E, A0], error: E): IO[Unit, A0]
}

object Supervisor {
  final def none: Supervisor[Any] = retry(Schedule.never)

  final def retry[E, A](policy: Schedule[E, A]): Supervisor[E] =
    retryOrElse(policy, (e: E, _: A) => ZIO.fail(e))

  final def retryOrElse[E, A](policy: Schedule[E, A], orElse: (E, A) => IO[E, A]): Supervisor[E] =
    new Supervisor[E] {
      override type A0 = A

      override def supervise(io: IO[E, A0], error: E): IO[Unit, A0] = 
        io.retryOrElse(policy, orElse).mapError(_ => ()).provide(Clock.Live)
    }
}
```
However, this fails to compile with the following (expected) error:

```
[error] /home/dejan/Work/zio-actors/src/main/scala/zio/actors/Actor.scala:35:30: type mismatch;
[error]  found   : scalaz.zio.IO[E,(S, A)]
[error]     (which expands to)  scalaz.zio.ZIO[Any,E,(S, A)]
[error]  required: scalaz.zio.IO[E,supervisor.A0]
[error]     (which expands to)  scalaz.zio.ZIO[Any,E,supervisor.A0]
[error]                   .supervise(receiver, e)
[error]                              ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 0 s, completed May 2, 2019 11:21:34 PM
```

TLDR: in order to go with `retryOrElse` I can't simply use `_` as it was used in original supervisor. Path-dependent approach seemed to be the "dead end" as well. Any suggestions?

**Comment by jdegoes** [comment]
Existentials probably won't help here, since the supervisor must be polymorphic in the type of message.

What trouble do you find with the more direct encoding:

```scala
trait Supervisor[-E] {
  def supervise[A](io: IO[E, A], error: E): IO[Unit, A]
}

object Supervisor {
  final def none: Supervisor[Any] = retry(Schedule.never)

  final def retry[E, A](policy: Schedule[E, A]): Supervisor[E] = 
    new Supervisor[E] {
        def supervise[A](io: IO[E, A], error: E): IO[Unit, A] =
          io.retry(policy).mapError(_ => ()).provide(Clock.Live)
    }

  final def retryOrElse[E, A](policy: Schedule[E, A], orElse: (E, A) => IO[E, A]): Supervisor[E] =
    new Supervisor[E] {
      override def supervise(io: IO[E, A], error: E): IO[Unit, A] = 
          io.retryOrElse(policy, orElse).mapError(_ => ()).provide(Clock.Live)
    }
}
```

**Comment by mijicd** [comment]
Tried that one as well, but the compilation fails with `private method supervise is never used`, as it lacks `[A]` in `retryOrElse` implementations.

**Comment by jdegoes** [comment]
Ah, I see the problem. 

The naive solution is here:

```scala
trait Supervisor[-E] {
  def supervise[A](io: IO[E, A], error: E): IO[Unit, A]
}

object Supervisor {
  trait Fallback[-E, -A] {
    def orElse[A0](E, A): IO[E, A0]
  }

  final def none: Supervisor[Any] = retry(Schedule.never)

  final def retry[E, A](policy: Schedule[E, A]): Supervisor[E] = 
    new Supervisor[E] {
        def supervise[A](io: IO[E, A], error: E): IO[Unit, A] =
          io.retry(policy).mapError(_ => ()).provide(Clock.Live)
    }

  final def retryOrElse[E, A](policy: Schedule[E, A], fallback: Fallback[E, A]): Supervisor[E] =
    new Supervisor[E] {
      override def supervise[A0](io: IO[E, A0], error: E): IO[Unit, A0] = 
          io.retryOrElse(policy, fallback.orElse).mapError(_ => ()).provide(Clock.Live)
    }
}
```

However, this shows the `orElse` is not that useful, because it must always fail. Perhaps instead of fallback, we could just allow an effect:

```scala
trait Supervisor[-E] {
  def supervise[A](io: IO[E, A], error: E): IO[Unit, A]
}

object Supervisor {
  final def none: Supervisor[Any] = retry(Schedule.never)

  final def retry[E, A](policy: Schedule[E, A]): Supervisor[E] = 
    new Supervisor[E] {
        def supervise[A](io: IO[E, A], error: E): IO[Unit, A] =
          io.retry(policy).mapError(_ => ()).provide(Clock.Live)
    }

  final def retryOrElse[E, A, B](policy: Schedule[E, A], orElse: IO[E, Unit]): Supervisor[E] =
    new Supervisor[E] {
      override def supervise[A0](io: IO[E, A0], error: E): IO[Unit, A0] = 
          io.retryOrElse(policy, (e, a) => orElse(e, a) *> ZIO.fail(error)).mapError(_ => ()).provide(Clock.Live)
    }
}
```

**Comment by mijicd** [comment]
The latter checks with `retryOrElse` slightly modified:

```scala
final def retryOrElse[E, A](policy: Schedule[E, A], orElse: IO[E, Unit]): Supervisor[E] =
  new Supervisor[E] {
    override def supervise[A0](io: IO[E, A0], error: E): IO[Unit, A0] =
      io.retryOrElse(policy, (e, a) => orElse(e, a) *> ZIO.fail(error)).mapError(_ => ()).provide(Clock.Live)
  }
```

Does it make sense to implement `retry` in terms of `retryOrElse` (as its done in ZIO):

```scala
final def retry[E, A](policy: Schedule[E, A]): Supervisor[E] =
  retryOrElse(policy, (e: E, _: A) => ZIO.fail(e))
```

**Comment by jdegoes** [comment]
Yes, it makes sense, in fact the `orElse` could just be `(_, _) => IO.unit`! 👍
